src/hotspot/share/services/mallocTracker.hpp

```diff
@@ -191,10 +191,18 @@ class MallocMemorySnapshot : public ResourceObj {
     // copy is going on, because their size is adjusted using this
     // buffer in make_adjustment().
     ThreadCritical tc;
-    s->_all_mallocs = _all_mallocs;
-    for (int index = 0; index < mt_number_of_types; index ++) {
-      s->_malloc[index] = _malloc[index];
-    }
+    size_t total_size;
+    size_t loop_counter = 0;
+    const size_t loop_limit = 100;
+    do {
+      total_size = 0;
+      s->_all_mallocs = _all_mallocs;
+      for (int index = 0; index < mt_number_of_types; index ++) {
+        s->_malloc[index] = _malloc[index];
+        total_size += _malloc[index].malloc_size();
+      }
+    } while(s->_all_mallocs.size() != total_size && ++loop_counter < loop_limit);
+    assert(s->_all_mallocs.size() == total_size, "Total != sum of parts");
```

Comment:
Preexisting, but could you please move this to the cpp file? It should not live in the header.

Also I'm not sure the assert is really useful. Won't that be just a source for very rare intermittent errors? What would be the action we take if that assert fires, increase the loop count?

Reply:
We copy `MemoryCounter` by value, not reference:

`s->_all_mallocs = _all_mallocs;`

where `_all_mallocs` is `MemoryCounter`, not  `*MemoryCounter`, so that assert can never trigger, right?

I mean, even if more memory gets allocated between when we exit the loop and assert, `_all_mallocs.size()` might change, but `s->_all_mallocs.size()` is frozen in time, isn't it?

Reply:
My understanding is that we make a snapshot of 2 moving parts: `_all_mallocs` and `_malloc[nmt_type]`

The problem is that after we snapshot `_all_mallocs`, `_malloc[nmt_type]` can change, so that:

`s._all_mallocs.size() != s._malloc[nmt_type1] + s._malloc[nmt_type2] + ...`

so we keep trying in a loop until we catch the system in a quiet enough period where:

`s._all_mallocs.size() == s._malloc[nmt_type1] + s._malloc[nmt_type2] + ...`

so when we come out of the loop then the **assert** is guaranteed never to trigger, because we made copies that are not going to change.

Reply:
> so we keep trying in a loop until we catch the system in a quiet enough period where:
> 
> `s._all_mallocs.size() == s._malloc[nmt_type1] + s._malloc[nmt_type2] + ...`

Either that or until the loop count is exhausted. That's what I meant - the assert could only fire if 100 tries were not enough to get a consistent picture.

Reply:
I see now what you meant. On one hand it would be nice to know if the loop was unable to get consistent copy, on the other hand if we do get an assert, the only thing we can do is increase the count.